smartsim/_core/mli/client/protoclient.py

```diff
@@ -0,0 +1,348 @@
+        to a pre-existing backbone featurestore. Requires the
+        environment to contain `_SMARTSIM_INFRA_BACKBONE`
+        environment variable.
+
+        :returns: The attached backbone featurestore
+        :raises SmartSimError: If the backbone descriptor is not contained
+         in the appropriate environment variable
+        """
+        descriptor = os.environ.get(BackboneFeatureStore.MLI_BACKBONE, None)
+        if descriptor is None or not descriptor:
+            raise SmartSimError(
+                "Missing required backbone configuration in environment: "
+                f"{BackboneFeatureStore.MLI_BACKBONE}"
+            )
+
+        backbone = t.cast(
+            BackboneFeatureStore, BackboneFeatureStore.from_descriptor(descriptor)
+        )
+        return backbone
+
+    def _attach_to_worker_queue(self) -> DragonFLIChannel:
+        """Wait until the backbone contains the worker queue configuration,
+        then attach an FLI to the given worker queue.
+
+        :returns: The attached FLI channel
+        :raises SmartSimError: if the required configuration is not found in the
+        backbone feature store
+        """
+
+        descriptor = ""
+        try:
+            # NOTE: without wait_for, this MUST be in the backbone....
+            config = self._backbone.wait_for(
+                [BackboneFeatureStore.MLI_WORKER_QUEUE], self.backbone_timeout
+            )
+            descriptor = str(config[BackboneFeatureStore.MLI_WORKER_QUEUE])
+        except Exception as ex:
+            logger.info(
+                f"Unable to retrieve {BackboneFeatureStore.MLI_WORKER_QUEUE} "
+                "to attach to the worker queue."
+            )
+            raise SmartSimError("Unable to locate worker queue using backbone") from ex
+
+        return DragonFLIChannel.from_descriptor(descriptor)
+
+    def _create_broadcaster(self) -> EventBroadcaster:
+        """Create an EventBroadcaster that broadcasts events to
+        all MLI components registered to consume them.
+
+        :returns: An EventBroadcaster instance
+        """
+        broadcaster = EventBroadcaster(
+            self._backbone, DragonCommChannel.from_descriptor
+        )
+        return broadcaster
+
+    def __init__(
+        self,
+        timing_on: bool,
+        backbone_timeout: float = _DEFAULT_BACKBONE_TIMEOUT,
+    ) -> None:
+        """Initialize the client instance.
+
+        :param timing_on: Flag indicating if timing information should be
+         written to file
+        :param backbone_timeout: Maximum wait time (in seconds) allowed to attach to the
+         worker queue
+        :raises SmartSimError: If unable to attach to a backbone featurestore
+        :raises ValueError: If an invalid backbone timeout is specified
+        """
+        if MPI is not None:
+            # TODO: determine a way to make MPI work in the test environment
+            #  - consider catching the import exception and defaulting rank to 0
```

Comment:
Yeah, that should be the way. I don't think we'll keep MPI4PY around much longer, anyhow. It was mostly a mock-app-related choice, but the MPI-dependent logic could be extracted (we could create it with a rank argument, for example). We'll fix this.

Reply:
added link to this note in ticket for fixing.